gam/clustering.py

```diff
@@ -739,11 +877,126 @@ def _find_medoids(
+            # sample a batch from S_ref (for init, S_ref = X)
+            idx_ref = np.random.choice(
+                unselected_ids, size=self.batchsize, replace=True
+            )
+            ci_scale = math.sqrt(
+                (2 * math.log(1.0 / delta)) / (n_used_ref + self.batchsize)
+            )
+            outputs = da.apply_along_axis(
+                lambda j: self._dask_looping_solution_ids(
+                    np.array(X),
+                    idx_ref,
+                    dist_func,
+                    d_nearest,
+                    n_used_ref,
+                    mu_x,
+                    sigma_x,
+                    j,
+                    i,
+                ),
+                arr=solution_ids.reshape(-1, 1),
+                axis=1,
+            ).compute()
+            sigma_x, mu_x = outputs[:, 0], outputs[:, 1]
+
+            # Remove pts that are unlikely to be a solution
+            C_x = ci_scale * sigma_x
+            ucb = mu_x + C_x
+
+            # check if LCB of target is <= UCB of current best
+            lcb_target = mu_x - C_x
+            ucb_best = np.min(ucb)
+            solution_ids = np.where(lcb_target <= ucb_best)[0]
+
+            # clean up any center idx that crept in...
+            # for ic in centers:
+            #     if ic in solution_ids:
+            solution_ids = solution_ids[np.isin(solution_ids, centers, invert=True)]
+
+            n_used_ref = n_used_ref + self.batchsize
+
+        # finish search over the remaining candidates
+        if verbose:
+            print(
+                f"Final eval with candidates = {solution_ids.shape[0]}"
+            )  # , {solution_ids}")
+        if solution_ids.shape[0] == 1:
+            # save the single sample as a medoid
+            centers[i] = solution_ids  # probably a type error
+            d = (
+                cdist(X, X[centers[i], :].reshape(1, -1), metric=dist_func)
+                .squeeze()
+                .reshape(-1, 1)
+            )
+            d_best = np.array(d).reshape(-1, 1)
+        else:  # this is fastPam build - with far fewer pts to evaluate
+            tmp_arr = np.zeros((n_samples))
+            td = float("inf")
+            output_arr = da.apply_along_axis(
+                lambda j: self._dask_bandit_search_singles(
+                    np.array(X), dist_func, d_nearest, tmp_arr, j
+                ),
+                arr=da.array(solution_ids.reshape(-1, 1)),
+                axis=1,
+            ).compute()
+            idx = np.argmin(output_arr)
+            centers[i] = solution_ids[idx]
+            d_best = (
+                cdist(X, X[centers[i], :].reshape(1, -1), metric=dist_func)
+                .squeeze()
+                .reshape(-1, 1)
+            )
+        if i == [0]:
+            self.D = d_best
+        else:
+            self.D = np.concatenate((self.D, d_best), axis=1)
+        print("\t updated centers - ", centers)
```

Comment:
delete this too?

Reply:
that was a part of the initial code, it can be good to show what indices are being found as centers.